[bukuserver API] improve tag replacement/deletion

2ynn opened the pull request:
Addresses most comments made by @LeXofLeviafan  in [issue #672](https://github.com/jarun/buku/issues/672)

1. `ApiTagView.put()` (PUT `/api/tags/<tag>`) now expects a list of strings as arguments instead of a comma-separated string. Passing an empty list results in bad response 400.

2. A separate `ApiTagView.delete()` method was added to delete an existing tag (DELETE `/api/tags/<tag>`).

3. A trailing slash is now permitted on  `GET /api/tags/` route.


ORIGINAL DESCRIPTION
_0. Clarify the expected data type received in ApiTagView.put:
PUT /api/tags/<tag> now returns 400 if no string is passed as input data
1. Allow `ApiTagView.put()` to update an existing tag with multiple ones by passing a single, comma-separated string. Instead of modifying the behavior of `replace_tag()`, which in my opinion would come with too many side effects, I have decided to wrap the `new_tags` comma-separated string into a list, in the same way as it is done in the [TagModelView.update_model()](https://github.com/jarun/buku/blob/f00348e722f03ae710a3a9344e7ad83a4b0f2a42/bukuserver/views.py#L573) method.
2. Allow tag deletion through passing an empty string to ApiTagView.put.
Passing `{tags: ''}` as data to `PUT /api/tags/<tag>` now makes it possible to delete an existing tag.
The `delete_tag_at_index()` function is now called with `chatty=False` to bypass confirmation.  To do so, an additional optional argument `chatty` had to be added to the `replace_tag()` function.
3. Allow trailing slashes in URLs by setting `strict_slashes=False` to all api endpoints
4. Unrelated but `test_bookmark_range_api` was not passing on master due to bookmark titles not matching between parametrization and assertion._

LeXofLeviafan commented:
To clarify:

> Instead of modifying the behavior of `replace_tag()`, which in my opinion would come with too many side effects, I have decided to wrap the `new_tags` comma-separated string into a list

It's better to make the input format straightforward instead of working around a tricky one whenever you use the function. A simple "accepting a list & processing it as-is" is a much better behaviour for a function API than "accepting a list and mashing its contents together in a way that takes effort to describe".

> Passing `{tags: ''}` as data to PUT `/api/tags/<tag>` now makes it possible to delete an existing tag.

Since this API appears to be going for RESTful approach, it's probably better to make deletion a separate operation, instead of combining it with edit operation. (…The reasoning is the same here, come to think of it: web/library API doesn't have the same limitations as CLI so there's no need to complicate it by imposing them anyway.)

P.S. You may want to amend existing commits & force-push your changes instead of adding new commits; the repo owner prefers to avoid having multiple small commits (but he doesn't want to use squash-merge).

LeXofLeviafan commented on bukuserver/server.py:
```diff
@@ -127,6 +127,7 @@ def shell_context():
     )
     # routing
     #  api
+    app.url_map.strict_slashes = False
```
…I meant the `/api/tags` endpoint specifically tho

LeXofLeviafan commented on tests/test_server.py:
```diff
@@ -209,7 +213,7 @@ def test_bookmark_range_api(client):
         'bookmarks': {
             '1': {'description': '', 'tags': [], 'title': 'Google', 'url': 'http://google.com'},
-            '2': {'description': '', 'tags': [], 'title': 'Example Domain', 'url': 'http://example.com'}}}
+            '2': {'description': '', 'tags': [], 'title': None, 'url': 'http://example.com'}}}
```
…What's the purpose of this change?

2ynn replied:
`test_bookmark_range_api` does not pass on master due to bookmark titles not matching between parametrization and assertion.

jarun commented:
@LeXofLeviafan can you have  a look at the build failure on master?

LeXofLeviafan commented:
@jarun the only thing that fails to pass on current master is `pylint` (due to having `else` on `try` blocks with `catch` blocks ending on `return`s).
Maybe it should be moved into a separate check (along with `flake8`); I doubt that it'd be much different between Python versions, and on the result would be making clear whether it's linting or tests that doesn't pass.

jarun commented:
> pylint (due to having else on try blocks with catch blocks ending on returns)

Why can't we remove the else? What is this syntax `try ... else`? Why can't it be avoided?

2ynn commented:
> @2ynn try doing a rebase on `upstream/master` – the current master is passing all checks.

Hi @LeXofLeviafan, I checked out the `upstream/master` and I still get the same error on `test_bookmark_range_api` when I run `pytest -vv tests_server.py`: 
```
client = <FlaskClient <FlaskAPI 'bukuserver.server'>>

    def test_bookmark_range_api(client):
        status_code = 200
        kwargs_list = [
            {"data": {'url': 'http://google.com'}},
            {"data": {'url': 'http://example.com'}}]
        for kwargs in kwargs_list:
            rd = client.post('/api/bookmarks', **kwargs)
            assert rd.status_code == status_code
        rd = client.get('/api/bookmarks/1/2')
        assert rd.status_code == status_code
>       assert rd.get_json() == {
            'bookmarks': {
                '1': {'description': '', 'tags': [], 'title': 'Google', 'url': 'http://google.com'},
                '2': {'description': '', 'tags': [], 'title': 'Example Domain', 'url': 'http://example.com'}}}
E       AssertionError: assert {'bookmarks': {'1': {'description': '', 'tags': [], 'title': 'Google', 'url': 'http://google.com'}, '2': {'description': '', 'tags': [], 'title': None, 'url': 'http://example.com'}}} == {'bookmarks': {'1': {'description': '', 'tags': [], 'title': 'Google', 'url': 'http://google.com'}, '2': {'description': '', 'tags': [], 'title': 'Example Domain', 'url': 'http://example.com'}}}
E         Differing items:
E         {'bookmarks': {'1': {'description': '', 'tags': [], 'title': 'Google', 'url': 'http://google.com'}, '2': {'description': '', 'tags': [], 'title': None, 'url': 'http://example.com'}}} != {'bookmarks': {'1': {'description': '', 'tags': [], 'title': 'Google', 'url': 'http://google.com'}, '2': {'description': '', 'tags': [], 'title': 'Example Domain', 'url': 'http://example.com'}}}
E         Full diff:
E           {
E            'bookmarks': {'1': {'description': '',
E                                'tags': [],
E                                'title': 'Google',
E                                'url': 'http://google.com'},
E                          '2': {'description': '',
E                                'tags': [],
E         -                      'title': 'Example Domain',
E         ?                               ^^^^^^^ --------
E         +                      'title': None,
E         ?                               ^^^
E                                'url': 'http://example.com'}},
E           }

tests/test_server.py:209: AssertionError
```

LeXofLeviafan commented:
…I've looked it over, and found that the cause here is likely twofold:
* the test is intended specifically for the web API range query, but it uses other web API operations (instead of doing environment setup directly); the test fails because the _db setup section_ produces unexpected result
* the default behaviour for the "add bookmark" operation is "fetch missing data from the Internet" – not something you'd want done by default _in the code_, and _especially_ in the tests (in the code you'd want to activate this _explicitly_, and in a test you'd want to mock every such network request _if you need it it at all_); the reason the test fails on your machine is because it's doing something that no test should really ever do: interact with entities outside the test environment (live network, in this case)

In other words, what should be changed isn't the value check but the db setup section.

(…As for exact reason why you're not getting back a webpage when fetching `http://example.com`, I couldn't tell, but it shouldn't be something that _matters_ when running a test.)

jarun commented on buku:
```diff
@@ -1970,23 +1970,24 @@ class BukuDb:
-        if newtags == DELIM:
-            return self.delete_tag_at_index(0, orig)
+        # Remove original tag from DB if new tagset is empty
+        if newtags == '':
```
Does Example 31 in README `$ buku --replace 'old tag'` work with this anymore?

jarun replied:
Also, Example 30.

LeXofLeviafan replied:
P.S. A pythonic way of checking for the string being empty would be `if not newtags`, BTW

jarun replied:
So what would be the syntax to replace a tag by 2 tags from the command line?

LeXofLeviafan replied:
Same as now, I believe? `--replace foo bar,baz`
My suggestions are meant to keep the old CLI behaviour working.

…Alternatively, the CLI syntax can be changed to treat separate parameters as separate tags – in this case, tags containing whitespace would need to be quoted explicitly (i.e. `'foo bar'`  or `"foo bar"`); that's how it's normally done with shell arguments. In this case, `args.replace[1:]` can be passed directly (without preprocessing).

jarun replied:
Yes, please keep the behaviour same as now.

LeXofLeviafan replied:
@2ynn the code should be updated to retain existing CLI behaviour (as mentioned [above](https://github.com/jarun/buku/pull/676#discussion_r1130310495)).

To rephrase, there's two options:
1. Do `','.join(new).split(',')` before [processing input](https://github.com/2ynn/buku/blob/put_api_tags/buku#L1982), and do `parse_tags(args.replace[:1])` [in the CLI](https://github.com/2ynn/buku/blob/put_api_tags/buku#L5985)
2. _Replace_ input processing with `parse_tags(','.join(new))`, and do `[' '.join(args.replace[1:])]` in the CLI

The latter should result in simpler code, I believe (though the ‘empty’ value would be `DELIM` again).
…If you pick the former instead, make sure to strip whitespace and filter out empty tag values as well, these.

2ynn replied:
My initial approach was to no alter buku CLI behavior. I personally believe that if `replace_tag()` is intended in that way it should accept a comma-separated string as an argument and not a list. I have modified my code on the API side to parse the input list and create a separated string to be passed to `replace_tags()`. For now this string is wrapped in list as a hack. What is the type of `args.replace[1:]` ?

LeXofLeviafan replied:
You seem to misunderstand. I didn't mean you should _reverse_ the changes in the function: if it accepts a list, it _should_ treat separate items as such (instead of implicitly joining them with spaces and re-splitting with commas). Alternatively this argument could be made a string, to do away with ambiguity altogether (it'll be split over commas either way).

The problem with your change was that you didn't _update_ CLI code so that it works the same way as it did before. The CLI code accepts a _comma-separated_ tags list which is _split over spaces_ (to avoid needing quotes, I guess?), and passes it into the function _unprocessed_; after changing how the function handles its argument, you need to update the code that _uses_ it.

Like I said before, the simplest way of handling this is to move the `[string]` hack inside of the function – then the CLI-specific space-joining can be done in the CLI code, and any other place where it's called won't have to deal with this CLI-specific stuff (whether the argument is kept as a list of tags, or made a comma-joined string).

2ynn replied:
Ok, thank you for the clarifications. I am going with option 2 but `parse_tags()` is intended to take a list as an argument, passing it a string results in treating the input as a list of characters and adds spaces between each of them. Did you mean `parse_tags([DELIM.join(new)])` ?

LeXofLeviafan replied:
Yes, basically.

…You don't really need to create a variable just to pass its value into a call immediately, BTW. Especially with that short of an expression.

LeXofLeviafan replied:
> `[tag.strip(buku.DELIM) for tag in new_tags]`

This doesn't seem necessary, as the commas are handled within the function.

2ynn replied:
from an API perspective, if a user wants to replace an existing tag with tags that happen to have commas in them (e.g. `['tag,2', 'tag,3']) wouldn't the result be unexpected ( `tag`, `1` and `2` in that case) ?

LeXofLeviafan replied:
* `.strip()` _does not affect the middle of the string_: `',foo,bar,baz,'.strip(',') == 'foo,bar,baz'`
* the result of `parse_tags([','.join([',foo,', ',bar,'])])` is exactly the same as `parse_tags(['foo,bar'])`
* commas are not supported in tags: they're stored as a single string in DB, with `,` as separator
* this is the reason why I suggested making this argument into a string as an alternative solution

If you want to avoid surprises for the user, you can reject requests containing invalid tags. (…Incidentally, have you checked what happens if the original tag argument contains commas? It doesn't seem to be covered by tests.)

> `tags == []`

The pythonic way of checking for emptiness is via `not`, BTW.

LeXofLeviafan replied:
…A newline is necessary to end a quotation block in Markdown :sweat_smile:

> OK, I confused it with `replace(',', ' ')`

Wouldn't that be even more unexpected?

> I also think returning more detailed error strings than "failure" would be useful.

I suppose you could introduce an optional argument for that. That would require converting values in `response_template` into lambdas.

…Incidentally, `response_ok` and `response_bad` might as well be moved into `response.py` (thus changing their invocations to `response.ok` and `response.bad`).

>  Could this case be covered by a 404 on a non-existing tag or would passing a string containing commas to `search_tag` be problematic?

I just checked and it seems that the current implementation [doesn't account for it at all](https://github.com/jarun/buku/issues/680). I suppose we need to figure out how it _should_ be handled first, and then adjust API behaviour accordingly (…and fix the implementation as well :sweat_smile:).

LeXofLeviafan replied:
…Speaking of which, the `DELETE` method also should be checked for the commas handling.

LeXofLeviafan replied:
…Based on the issue discussion, it seems that attempts to replace multiple tags at once should be rejected. I suggest implementing in `replace_tag()` a check for comma in the first argument (handling it the way it does the second argument being identical to the first).

As for the network API, I guess it should produce HTTP 404 for such URIs.

2ynn commented:
@jarun @LeXofLeviafan I'm really sorry for the silence, I should be able will to get back on it this week-end.

jarun commented:
No problem! Thanks guys!

LeXofLeviafan commented:
…Incidentally, the reason tests are failing [is not related to these changes](https://github.com/jarun/buku/issues/686).

2ynn commented on bukuserver/api.py:
```diff
@@ -101,16 +97,33 @@ def get(self, tag: T.Optional[str]):
     def put(self, tag: str):
+        try:
+            tags = request.data.getlist('tags')  # werkzeug.datastructures.MultiDict
+        except AttributeError:
```
When passing `data={'tags': ['tag1', 'tag2']}` using `client.put` in `test_server.py` the received object is of type `werkzeug.datastructures.MultiDict`:
```
request.data.getlist('tags') -> ['tag1', 'tag2']  (list)
request.get('tags') -> 'tag1' (str)
```
However, when passing `data={'tags':['tag1', 'tag2']} using `curl`:
```
request.data.get('tags') -> ['tag1', 'tag2']  (list)
request.data.getlist('tags') -> AttributeError
```

Note that as a side effect ` rd = client.put('/api/tags/old', data={'tags': 'new'})` will return 200 whereas passing the same data with curl will fail with 400...

LeXofLeviafan replied:
Ah, that's cuz you're _not_ actually submitting JSON with `client.put()` unless you specify it as such, e.g. by [passing the Content-Type header](https://github.com/jarun/buku/blob/v4.8/tests/test_server.py#L214-L215) (what you're sending here instead is `application/x-www-form-urlencoded` with body `tags=tag1&tags=tag2`, which results in a different parser being applied).

I suspect that only JSON is meant to be supported by the API, and thus the tests should be refactored to actually use it (although rather than via custom HTTP headers, I suggest [replacing `data=` with `json=`](https://werkzeug.palletsprojects.com/en/2.2.x/test/#request-body)).

LeXofLeviafan commented on bukuserver/api.py:
```diff
@@ -91,6 +84,9 @@ def search_tag(
 class ApiTagView(MethodView):
 
+    def is_valid_tag(self, tag: str):
+        return isinstance(tag, str) and buku.DELIM not in tag
```
…What I meant here is that _this function_ doesn't need to be a _method_ (it's a _utility_ that has no need to know that `ApiTagView` even _exists_, let alone having access to the state of the current instance) :sweat_smile:

LeXofLeviafan commented:
BTW, [this](https://github.com/jarun/buku/blob/v4.8/tests/test_server.py#L124) seems like something that should be updated as well in this pull-request (for sake of consistency, at least).

Also, for some reason, [this test](https://github.com/jarun/buku/blob/v4.8/tests/test_bukuDb.py#L974) seems to be [failing](https://github.com/jarun/buku/blob/v4.8/tests/test_bukuDb.py#L998) for me on your branch, when running tests locally. Not sure why; can you reproduce? (…Incidentally, I've noticed that [here](https://github.com/jarun/buku/blob/v4.8/tests/test_bukuDb.py#L994), a boolean is compared to a string literal; may want to change that too, I reckon :sweat_smile:)

2ynn commented:
> Also, for some reason, [this test](https://github.com/jarun/buku/blob/v4.8/tests/test_bukuDb.py#L974) seems to be [failing](https://github.com/jarun/buku/blob/v4.8/tests/test_bukuDb.py#L998) for me on your branch, when running tests locally. Not sure why; can you reproduce?

Yes this test also fails for me, before and after changing [this line](https://github.com/jarun/buku/blob/v4.8/tests/test_bukuDb.py#L994). However on my local machine I get 27failed, 90 passed when running `test_bukuDb.py` . . .

> BTW, [this](https://github.com/jarun/buku/blob/v4.8/tests/test_server.py#L124) seems like something that should be updated as well in this pull-request (for sake of consistency, at least).
I decided to switch to `wtform_json` to allow passing json data while using proper form validation. (I will squash my commits if you agree with the changes).

The new lint error is due to FlaskForm having no `from_json` method by default. I believe `wtforms_json.init()` is responsible for adding these methods in `create_app()` so I don't how to fix this error.  Note  `tox -e pylint` doesn't complain locally.

LeXofLeviafan commented:
> Yes this test also fails for me, before and after changing [this line](https://github.com/jarun/buku/blob/v4.8/tests/test_bukuDb.py#L994).

After changing this line, it still fails on case `1-True-False` (`index=1`, `delay_commit=True`, `input_retval=False`).

> The new lint error is due to FlaskForm having no from_json() method by default.

You could at least [disable the check](https://pylint.readthedocs.io/en/latest/user_guide/messages/message_control.html#block-disables) on those lines.

LeXofLeviafan commented on setup.py:
```diff
@@ -41,6 +41,7 @@
     "flask-paginate>=2022.1.8",
     "Flask-WTF>=1.0.1",
     "Flask>=2.2.2,<2.3",
+    "WTForms-JSON>=0.3.5",
```
The file `bukuserver/requirements.txt` must be kept in sync with `setup.py`.

2ynn replied:
i didn't have the impression it was kept in sync because of the comment on top:
`# use setup.py for latest required package `

LeXofLeviafan replied:
[There's a test for it.](https://github.com/jarun/buku/blob/v4.8/tests/test_setup.py#L17-L22)

LeXofLeviafan commented on bukuserver/server.py:
```diff
@@ -109,6 +110,8 @@ def create_app(db_file=None):
     app.app_context().push()
     setattr(flask.g, 'bukudb', bukudb)
     init_locale(app)
+    app.config["WTF_CSRF_ENABLED"] = False
```
Don't turn off security in production code.

If this package doesn't work with WTF_CSRF, you can use other means for data validation (like [JSON Schema](https://pypi.org/project/jsonschema)).

LeXofLeviafan commented on tests/test_server.py:
```diff
@@ -133,19 +133,24 @@ def test_bookmark_api(client):
         'description': '', 'tags': [], 'title': 'Google', 'url': 'http://google.com'}
-    rd = client.put('/api/bookmarks/1', data={'tags': [',tag1,tag2,']})
+    rd = client.put('/api/bookmarks/1', json={'tags': ['tag1', 'tag2']})
+    assert rd.status_code == 400
```
This request shouldn't fail tho.

…You do realize that POST and PUT requests should have different validation rules, right?

LeXofLeviafan replied:
REST isn't a strict guideline; there are variations between different implementations.

This particular API has no use for full-replacement-updates (and frankly speaking they're rarely needed in the first place), so the PUT method is used for partial updates instead (which is fairly common BTW); you could tell that by the previous test, if anything.

LeXofLeviafan commented:
> Note `tox -e pylint` doesn't complain locally.

There shouldn't be any discrepancies in behaviour; does the local `tox`/`pylint` get different config smh?…

LeXofLeviafan commented on tests/test_server.py:
```diff
@@ -222,9 +227,7 @@ def test_bookmark_range_api(client):
-    put_data = json.dumps({1: {'tags': 'tag1'}, 2: {'tags': 'tag2'}})
-    headers = {'content-type': 'application/json'}
-    rd = client.put('/api/bookmarks/1/2', data=put_data, headers=headers)
+    rd = client.put('/api/bookmarks/1/2', json={1: {'tags': 'tag1'}, 2: {'tags': 'tag2'}})
```
…Should `tags` really be a _string_ (as in, not a _list_)?

LeXofLeviafan commented on bukuserver/forms.py:
```diff
@@ -1,26 +1,49 @@
+    if DELIM in field.data:
+        raise ValidationError(
+            "Tag must not contain delimiter ({})".format(DELIM)
+        )
```
What's with the multiline call?

2ynn replied:
Are you referring to the last comment for `test_bookmark_range_api`? I will change strings for lists of strings so I this check should still apply

LeXofLeviafan replied:
I'm saying there's no reason for this call to be multiline. This comment is attached to specific place in the code, you can see what I'm referring to.

LeXofLeviafan commented on bukuserver/forms.py:
```diff
@@ -1,26 +1,49 @@
+    fetch = HiddenField(filters=[bool])
+
+
+def delim_not_in_tag(form, field):
```
`valid_tag`?

2ynn replied:
I think a valid tag would be one that checks all validators.

LeXofLeviafan replied:
A "valid tag" is a _tag_ value which is _valid_. As of now, this is defined as "not containing a comma"; but "valid tag" is a higher-level concept and thus should be preferred to "not containing a comma" when identifying tag validation.

2ynn replied:
What i mean is that `delim_not_in_tag` is a validator method that checks for a specific condition related to delim, and raises a message accordingly. If criteria were to change in the future I think one could easily  add/remove other validators from the tags validators lists.
Btw I think it could also be stored in `validators.py` and called `delim_not_in_field` instead.  

Would you prefer something like `def custom_tag_validator(form, field)` instead?

LeXofLeviafan replied:
What I'm saying that you _shouldn't_ have a validator that "checks for a specific condition" (a low-level concept). This condition is only used for validating _tags_, but there's no good reason for any code outside of the validator itself to know what exact conditions a _tag_ must match to be _valid_.

What you should have is _one_ validator, that ensures validity of the value as a tag (a high-level concept), and being the only part of the code that is concerned with what exactly that constitutes.

As for naming, I'm following the general trend of validator naming here. Except this one isn't in PascalCase, I guess; I'm uncertain if it should be applied. (…Also, I'd say it should be moved to the top of the file, instead of randomly popping up somewhere in the middle.)

jarun commented:
@LeXofLeviafan please squash and merge when this is ready. Please ensure documentation (man, readme, wiki) is updated of required.

2ynn commented on tests/test_server.py:
```diff
@@ -58,9 +57,9 @@ def test_api_empty_db(client, url, exp_res):
+        ['/api/tags/1', None, 400, 'put'],
         ['/api/bookmarks/1', response_template['failure'], 400, 'get'],
-        ['/api/bookmarks/1', response_template['failure'], 400, 'put'],
+        ['/api/bookmarks/1', None, 400, 'put'],
```
I did notice too. I can investigate more.
As a comparison, if I  use curl without providing any data along with the PUT request,   I receive this response (html, not json):
```
curl -X PUT 'http://127.0.0.1:5000/api/bookmarks/1'

<!doctype html>
<html lang=en>
<title>400 Bad Request</title>
<h1>Bad Request</h1>
<p>Did not attempt to load JSON data because the request Content-Type was not &#39;application/json&#39;.</p>
```

2ynn replied:
> Come to think of it… why is it that in these cases alone, the response is None? There's a bunch of other HTTP 400 in the test cases, but all of those return data.

All the other cases provide the json parameter to `client.put. So we are not really checking for an invalid_id in the `post` and `put` cases.

2ynn replied:
The exception is thrown when calling `request.get_json()` while retrieving the form:
```
<class 'werkzeug.exceptions.BadRequest'>
Exception 400 Bad Request: The browser (or proxy) sent a request that this server could not understand.
 ```
I can catch it and return `response.bad()` instead.

In the `'/api/bookmarks/1/2', response_template['failure'], 400, 'put']` case the `reponse.bad()` was returned when checking for an invalid range, before trying to get the json.

LeXofLeviafan replied:
…Oddly enough this seems to be passing on py37 (but not py38, py39 nor py310).

2ynn commented on bukuserver/api.py:
```diff
@@ -101,16 +94,28 @@ def get(self, tag: T.Optional[str]):
+        try:
+            form = forms.ApiTagForm.from_json(request.get_json())  # pylint: disable=no-member
+        except Exception as e:
+            current_app.log_exception(e)
+            return response.bad()
```
I would rather return `response.bad(str(e))` but that means the `test_invalid_id` test will have to be parametrized differently to not check for the response message. Would you be open to only check for the status_code?
Btw are we ok with the fact that raised `expeptions.NotFound()` return no `status` in the response json?

LeXofLeviafan replied:
> to not check for the response message

Why shouldn't it check for the response message? I'm pretty sure you can use `response.bad()` (or a simplified equivalent thereof) to produce responses for test cases.

> Btw are we ok with the fact that raised expeptions.NotFound() (e.g. [line 100](https://github.com/jarun/buku/blob/db054e154dfe6cf26c47f92a8fa59504e39f1a48/bukuserver/api.py#L100) ) return no status in the response json?

I suppose the response format should stay consistent across the API…

2ynn replied:
I  was concerned that if I returned to whole exception message it might change with time or the implementation... But OK, as commented above I will return simplified messages and test against them. I will use some sort of template (or simply constants, e.g. `TAG_NOT_VALID = "Tag is not valid"`) that can be reused to avoid duplicating strings between response.py and test_server.py

LeXofLeviafan replied:
You can use an [enum](https://docs.python.org/3.10/library/enum.html), I guess (e.g. `Message` or `Error`)

LeXofLeviafan commented on tests/test_server.py:
```diff
@@ -86,17 +85,19 @@ def test_tag_api(client):
+    rd = client.put('/api/tags/tag1', json={'tags': []})
     assert rd.status_code == 400
-    rd = client.put('/api/tags/tag1', data={'tags': []})
+    rd = client.put('/api/tags/tag1', json={'tags': 'string'})
```
…Can we make `message` a string though? The client shouldn't be boggled down with specifics beyond "request failed, here's why".

I believe just showing the first error message should suffice (e.g. `Invalid 'tags' (This field is required)`).

(…Alternatively, there could be an additional metadata field for structural information, while `message` would be used for providing a simplified response.)

LeXofLeviafan commented on tests/test_server.py:
```diff
@@ -27,6 +34,7 @@ def test_cli(args, word):
 def client(tmp_path):
     test_db = tmp_path / 'test.db'
     app = server.create_app(test_db.as_posix())
+    app.config["WTF_CSRF_ENABLED"] = False
```
Do note that this currently causes the tests to fail.

2ynn replied:
I think this one is needed for tests and had passed earlier with this config. That said I did put back the same line by mistake in `server.py` as i was also testing routes with curl locally, i think it might be the issue.

LeXofLeviafan replied:
…Isn't the actual problem that CSRF should be [disabled for API forms (or requests) only](https://wtforms.readthedocs.io/en/3.0.x/meta/#wtforms.meta.DefaultMeta.csrf), rather than globally? The API is hardly usable with CSRF (as the API is meant for external use), but it shouldn't be disabled for non-API requests.

LeXofLeviafan commented:
Regression: trying to create a new record without tags (UI) causes an error.

2ynn commented on tests/test_server.py:
```diff
@@ -1,14 +1,21 @@
+    assert rd.get_json() == exp_res.get_json()
+    assert rd.status_code == exp_res.get_status_code()
+
+def assert_result(rd, exp_dict :dict, status_code :int = 200):
```
I now use the same `assert_response` check everywhere as all responses are built using the `Response` Enum by passing optional data to update the response's json. Note that as a consequence all responses will contain a `status` and `message` fields, e.g. GET /api/tags/tag1 will return:
```
{'status': 0, 'message': 'Success.', 'name': 'tag1', usage_count': 1}
```
Let me know if that is an issue. I could also  wrap data into a separate dictionary, e.g.:
```
{'status': 0, 'message': 'Success.', 'data': {'name': 'tag1', usage_count': 1}}
```

2ynn commented:
> Regression: trying to create a new record without tags (UI) causes an error.

My bad, this was due to the `DataRequired()` validator in the `ApiBookmarkCreateForm.tags` field. 
Now both `json={'url': url, 'tags': [] }` and `json={'url': url}` work with post/put on `/api/bookmarks`

LeXofLeviafan commented:
> > Regression: trying to create a new record without tags (UI) causes an error.
> 
> My bad, this was due to the `DataRequired()` validator in the `ApiBookmarkCreateForm.tags` field. Now both `json={'url': url, 'tags': [] }` and `json={'url': url}` work with post/put on `/api/bookmarks`

I'm talking about **UI** here (not the API).

LeXofLeviafan commented on bukuserver/response.py:
```diff
@@ -1,4 +1,40 @@
+                self.get_status_code(), {'ContentType': 'application/json'})
+
+
+response_from_flag = lambda flag: Response.SUCCESS() if flag else Response.FAILURE()
```
This would probably make more sense as a [`@staticmethod`](https://docs.python.org/3.7/library/functions.html#staticmethod) of `Response` instead, usage wise…

LeXofLeviafan commented on bukuserver/api.py:
```diff
@@ -121,34 +132,50 @@ def get(self, rec_id: T.Union[int, None]):
+                return Response.TAGS_NOT_LIST()
+            form = ApiBookmarkCreateForm.from_json(json)  # pylint: disable=no-member
+        except BadRequest:
+            return Response.JSON_NOT_VALID
```
…Is this the intended return value?

LeXofLeviafan commented:
@2ynn You don't seem to have done anything about HTTP 415 in the failing tests for the last few commits ([even though you seem to have confirmed the discussed solution](https://github.com/jarun/buku/pull/676#discussion_r1186154077))… Did you forget about it? :sweat_smile:

2ynn commented:
> > > Regression: trying to create a new record without tags (UI) causes an error.
> > 
> > 
> > My bad, this was due to the `DataRequired()` validator in the `ApiBookmarkCreateForm.tags` field. Now both `json={'url': url, 'tags': [] }` and `json={'url': url}` work with post/put on `/api/bookmarks`
> 
> I'm talking about **UI** here (not the API).

Ok, in fact creating a new record failed even when providing tags. Issue disappeared in UI when i deleted the line `wtforms_json.init()`. I therefore removed the `wtforms-json` dependency by processing json data in two steps, e.g.
```
form = ApiBookmarkEditForm({})
form.process(data=data)
```
Note that passing `data` (or `ImmutableMultiDict(data)`) during form creation results in `form.tags.data = [None]`, causing form validation to fail.

https://wtforms.readthedocs.io/en/2.3.x/forms/#wtforms.form.BaseForm.process

2ynn commented on tests/test_server.py:
```diff
@@ -38,204 +53,206 @@ def test_home(client):
+@pytest.mark.parametrize(
+    'method, url, json, exp_res', [
+        ['get', '/api/tags/tag1', None, Response.TAG_NOT_FOUND],
+        ['put', '/api/tags/tag1', {'tags': ['tag2']}, Response.SUCCESS],
```
I am still feeling a bit uneasy about returning SUCCESS when replacing an non-existing tag (i.e. successfully replacing 0 occurrences of a tag). I feel like returning the number of records that were updated  in the response json would be better.

LeXofLeviafan replied:
The library function doesn't return that information though… May need to get @jarun to approve changing the code API again for that.

Or maybe just return 404 for non-existent tags instead?

2ynn replied:
404 works for me (as suggested in Issue #672).  I would rather open a separate PR if @jarun  later agrees to return the number of updated records.

LeXofLeviafan commented on bukuserver/response.py:
```diff
@@ -1,4 +1,45 @@
+
+    @property
+    def exit_code(self):
+        return int(self.status_code != status.HTTP_200_OK)
```
I'd say that the "boolean is implemented as `0`/`1` alias" thing is an implementation detail of CPython, and shouldn't be relied on in application logic.

It's better to define something like `OK, FAIL = 1, 0` at the top of the file and return those values explicitly by name (e.g. via `if`/`else`).

…Also, "exit code" is a process execution term – it hardly fits here as there's nothing being "exited" (and this is basically an oddly presented boolean, so calling it a "code" is also kind of a stretch). Maybe call it just `status` instead? (Thus making it so that "status" is `1` or `0`, and "status code" is `200`, `400`, `404`…)

LeXofLeviafan commented on bukuserver/response.py:
```diff
@@ -1,4 +1,45 @@
+
+    @staticmethod
+    def from_flag(flag: bool):
+        return Response.SUCCESS() if flag else Response.FAILURE()
```
I suggest moving this higher, next to the rest of the static members.

LeXofLeviafan commented on buku:
```diff
@@ -2002,8 +2006,6 @@ class BukuDb:
 
```
…Come to think of it – should this be actually printed out when not in CLI mode?

LeXofLeviafan commented on tests/test_server.py:
```diff
@@ -38,204 +53,206 @@ def test_home(client):
-    assert rd.get_json() == exp_res
+    assert rd.status_code in [
+        Response.JSON_NOT_VALID.status_code,
+        HTTP_415_UNSUPPORTED_MEDIA_TYPE
```
…I'm pretty sure 415 means the request wasn't submitted correctly (i.e. it's expecting JSON data in the body but instead it's plain text and empty) and thus never got processed.
If this pops up during a test, shouldn't the test code be fixed instead of saying it's the expected result?

2ynn replied:
Ok. It used to be (at least on my local machine) that providing no json would throw a `BadRequest` exception, thus returning JSON_NOT_VALID response.  However this is no longer the case since I stopped using `wtforms_json`.  I removed this test batch and added similar tests (with `json={}` instead) in `test_tag_api` and `test_bookmark_api`

2ynn commented:
@LeXofLeviafan sorry again for the delay and thank you so much for bearing with me, really appreciate your feedback.

I just realized that `bukudb.update_rec()` (used in PUT `/api/bookmarks/<rec_id>` and PUT `/api/bookmarks/<start>/<end>`) allows for `+` / `-` [prefixes to append/delete current tags](https://github.com/jarun/buku/blob/db054e154dfe6cf26c47f92a8fa59504e39f1a48/buku#L842)

I think this is a problem as we are not filtering such characters when receiving a list of tags. I would suggest adding a validator to `ApiBookmarkEditForm.tags` to reject any tag containing these characters.

2ynn commented on bukuserver/api.py:
```diff
@@ -158,45 +181,48 @@ def delete(self, rec_id: T.Union[int, None]):
+                    rec_id,
+                    form.url.data,
+                    form.title.data,
+                    None if data.get('tags') is None else buku.parse_tags([buku.DELIM.join(form.tags.data)]),
```
Sorry for my last (rushed) comment. My confusion came from the fact I was not using `parse_tags` within `update_rec`
Note here that if no `tags` field is provided in the request json, `form.tags.data` will resolve to `[]`, thus deleting all previous tags associated to the record, so I need to differentiate `json = {}` from `json = {'tags':[]}`
Would you prefer if I break this long parameter into multiple lines above the `update_rec()` call?

2ynn replied:
I mean that checking `form.tags.data` alone is not enough. For example if I only want to update the description (`json = {'description' = "foo"}`) then `tags_in=None` needs to be passed to `update_rec()`, not `form.tags.data = []`, otherwise all previous tags will be lost.

The PUT method performs partial update regarding form fields but when it comes to the `tags` field the whole list of tags must be provided, i.e. to add "tag3" to existing tags "tag1,tag2", `json={'tags':  ['tag1', 'tag2', 'tag3']} must be passed. Consequently a user can decide to remove all tags for a bookmark by passing `json={'tags': []}`, in which case an empty string is passed to `update_rec()` 

> I'm not actually sure that allowing to delete all tags for every record in the range is a good idea.

To delete all tags for every record in the range, empty lists would have to be passed for each index individually:
```
json={1: {'tags': []}, 2: {'tags': []}, ... )
```

Note that as it is now, the 'tags' string can be prepended with "+/-" to add/remove but the tags are directly passed to `update_rec` without performing all the checks that `parse_tags` would, so capital letters or duplicates could be passed for example.

LeXofLeviafan replied:
> To delete all tags for every record in the range, empty lists would have to be passed for each index individually:

My point is that mass-overwriting of tags probably shouldn't be of the "whole list replacement" kind in the first place; it's more likely to cause a problem (by destroying user data) rather than to solve one (since I rather suspect that wanting to discard all existing tags in a number of records to replace them with multiple identical tag sets regardless of previous content is not, in fact, a common usecase).

> when it comes to the tags field the whole list of tags must be provided, i.e. to add `tag3` to existing tags `tag1,tag2`, `json={'tags':  ['tag1', 'tag2', 'tag3']}` must be passed. Consequently a user can decide to remove all tags for a bookmark by passing `json={'tags': []}`, in which case an empty string is passed to `update_rec()`

And that's precisely why I'd say that having an option of passing `+tag3` would be much better, and _especially_ so when there's multiple _different_ tag lists to modify in such a way (and conversely, one could argue that _overwriting_ the tag list is permissible in case of editing one specific record, but allowing that in batch updates is basically just leaving a huge "mess up existing data" button waiting to be pressed by accident).

…Come to think of it, I can't think of a single usecase where allowing _any_ batch updates that write an _identical_ value to a field in multiple _different_ records would be helpful in the first place :thinking:
_All_ the field values in these bookmarks are specific per record, and stuff like "(un)marking all records from X to Y with a specific tag" also calls for a "modify the lists" operation, rather than "replace all lists with an identical one".

P.S. …Incidentally, what happens when both "modify" and "overwrite" tags (with/without `+`/`-`) are passed? Doesn't seem like this should be valid at all.

2ynn replied:
Note that `+tag3` is a valid tag and the `+` must be [prepended to the list of tags](https://github.com/jarun/buku/blob/master/buku#L878) when calling `update_rec()` (e.g `-,tag1` or `+,tag3,tag4`) so at the moment two consecutive PUT requests must be made to replace a tag  by another one.

Would you then consider allowing such input for `ApiBookmarkRangeView`: `{'tags': ['+,newtag', '-,oldtag', 'noeffect']}` and call `update_rec()` once per tag?

LeXofLeviafan replied:
> the `+` must be [prepended to the list of tags](https://github.com/jarun/buku/blob/master/buku#L878)

Is that how it works right now? Then, my P.S. doesn't apply after all, because it's checking the "value" of the first tag. (On the other hand, this must mean that `+` and `-` should not be valid tag names, since they're at least partially unsupported…)

> at the moment two consecutive PUT requests must be made to replace a tag by another one.  
> Would you then consider allowing such input for `ApiBookmarkRangeView`: `{'tags': ['+,newtag', '-,oldtag', 'noeffect']}` and call `update_rec()` once per tag?

`-,oldtag` has no effect when `oldtag` isn't present, but `+,newtag` is applied unconditionally; so "tag replacement" functionality is not supported, per se. (And the `noeffect` thing shouldn't really be valid in such circumstances.)

There's a few issues with implementing batch-batch tag modifications:
* batch edit is currently equivalent to "one edit per record", so allowing multiple edits per record goes outside of the current paradigm
* `tags` is meant to be specifically a list of tags; this makes it effectively a "list of lists" (which is confusingly expressed as a list of strings instead)
* this approach would require passing multiple-tags-as-one (which doesn't seem like a good thing when the API itself supports lists), and at the same time passing individual tags gets broken, smh
* such requests have potential to be able to freeze the server for long periods of time due to the lack a way to limit the amount of DB updates per record (unless you purposefully mesh all `+` and all `-` tag strings together, which would affect application order in case it matters)

2ynn replied:
>this must mean that + and - should not be valid tag names, since they're at least partially unsupported…
To clarify, '+'/'-' is prepended to the comma-separated list of tags (i.e. *before* the first comma), so I believe `+` and `-` can still be valid tags, e.g. `'+,+,-'` would add tags '+' and '-'

LeXofLeviafan commented on bukuserver/response.py:
```diff
@@ -1,4 +1,45 @@
+        json = {'status': self.status, 'message': self.message}
+        if isinstance(data, dict):
+            json.update(data)
+        return json
```
Technically this entire function is equivalent to the expression
```py
dict(status=self.status, message=self.message, **data or {})
```
(At least not accounting for the possibility of the data being an invalid value… but one could argue that a `TypeError` would be a more adequate result of an invalid invocation.)

…Also, naming it simply `.json()` might be better (especially since it's a pure function to begin with).

2ynn replied:
My initial intention with this function was to allow overriding the message or status  for specific cases without having to add new items to the Response Enum, e.g.:
```
Response.SUCCESS(data={'message':  "Successfully updated record"})
Response.FAILURE(data={'status':  status.HTTP_403_FORBIDDEN})
```
Although i concede it defeats the purpose of gathering all message strings in the same location. If it's not desired I will update using your solution.

LeXofLeviafan replied:
…Ah. I haven't considered the option of overriding the core fields.

Though TBH, making them arbitrary doesn't seem like a good idea to me; this file is a part of the web-app code, so if a new response is needed, it can simply be added here (and providing specifics is what metadata is _for_).

LeXofLeviafan replied:
…Incidentally, using a dict literal instead of `dict()` doesn't result in an error when passing existing keys (dict literals don't seem to have a check for repeating keys).

2ynn replied:
ok, I changed it to dict literal because pylint was complaining

LeXofLeviafan commented on bukuserver/response.py:
```diff
@@ -1,4 +1,45 @@
+        """
+        Kwargs
+           data (dict): updates the response's JSON
+        """
```
…If your goal was to have a single optional argument that must be explicitly named in the invocation, this might be a better way of doing it:
```py
def __call__ (self, *, data :dict = None):
```

LeXofLeviafan commented on tests/test_server.py:
```diff
@@ -38,204 +53,195 @@ def test_home(client):
-    rd = client.post('/api/bookmarks', data={'url': url})
-    assert rd.status_code == 200
-    assert rd.get_json() == response_template['success']
+    rd = client.post('/api/bookmarks', json={'url': url, 'tags': []})
```
Was adding a `tags` field necessary here? Is there a difference?

LeXofLeviafan commented:
@2ynn try running manual tests (or maybe [these ones](https://github.com/jarun/buku/pull/670) – from a local merge or the like) to see if anything changed; I haven't spotted any problems but it never hurts to double check.

Similarly, I haven't found any place in the docs that needs to be updated (the doc files only cover CLI usage as far as I can tell), but you may want to double check that as well.

LeXofLeviafan commented on bukuserver/api.py:
```diff
@@ -158,45 +185,52 @@ def delete(self, rec_id: T.Union[int, None]):
+            if data.get('tags') is None:
+                tags_str = None
+            else:
+                tags_str = buku.parse_tags([buku.DELIM.join(form.tags.data)])
```
…I'd say turning a fairly simple expression into a 4-line block is a bit of an overkill :sweat_smile:

LeXofLeviafan replied:
…Say. Does `form.tags.data` contain a value different from `data.get('tags')` when the latter is neither a string nor `None`? Because if not, it'd make sense to use the `tags` binding you added above in place of `form.tags.data` here.

LeXofLeviafan replied:
Come to think of it, if `data` is `None`, `data.get('tags')` would cause an `AttributeError` (and in the current version this call happens before form validation, so I'm guessing this would result in HTTP 500?).

2ynn replied:
yes if `form.validate()` returns True (i.e. there is no comma in strings), then `form.tags.data` should be the same as tags. However I pass form data for all other fields so i think it'd make sense to be consistent.

I suggest creating a [custom FieldList](https://wtforms.readthedocs.io/en/3.0.x/fields/#custom-fields) and let the form handle the processing from list to string

2ynn replied:
> Come to think of it, if `data` is `None`, `data.get('tags')` would cause an `AttributeError` (and in the current version this call happens before form validation, so I'm guessing this would result in HTTP 500?).

`data.get('tags') is short for data.get('tags', None) so it will return None if 'tags' field does not exist, not AttributeError

2ynn replied:
Sorry i misread you, you meant `data=None`. I think if the request if properly formed (i.e. json is passed) `request.get_json()` should not return None

LeXofLeviafan replied:
…And maybe the `TAGS_NOT_LIST` check should be actually checking for the `tags` value being a _list_ (as opposed to _string_)?

I also suspect that this is checked by the form validation.

…Speaking of which: shouldn't the request data be validated _before_ we start applying DB changes? As of right now, it seems that if the request is malformed, we end up applying some part of it before reporting a failure.

2ynn replied:
Yes the form is responsible for checking that the value associated with tags is a list. The reason why I first check that tags is not a string is because the form will still pass validation and treat the string value as a list of characters... For example when passing {`tags`: `not a list`} I get `tags_str = ,a,i,l,n,o,s,t`

True, I could validate all forms before proceeding.

LeXofLeviafan replied:
> `request.get_json()` should not return `None`

Technically `null` is a valid JSON that's equivalent to `None`. But the case I'm talking about is actually `get_json().get(…)` returning `None`.

LeXofLeviafan replied:
> I suggest creating a [custom FieldList](https://wtforms.readthedocs.io/en/3.0.x/fields/#custom-fields) and let the form handle the processing from list to string

That _is_ an option, I suppose… So long as it doesn't outright break on invalid values nor treat `None` as `[]` where it shouldn't.

2ynn replied:
So it turns out that `FieldList` was not actually checking that the value passed to `tags` was actually a list. What it does is it [loops through the data being passed](https://github.com/wtforms/wtforms/blob/master/src/wtforms/fields/list.py#L92) and fills its subfields (hence if data is a string it will loop through characters). Subfields are validated first, then FieldList validators are run (in this case there aren't any).  Adding a custom validator to `FieldList` did not help because the data was already processed when it reaches the validator (i.e. `None` resolves to `[]` and strings already split into a list of chars).
Instead I decided to create a custom `process_data` method in the form that processes the input, validates the form and parse the tags string for later use.

LeXofLeviafan replied:
…Make sure to double-check the API manually if you haven't yet (just in case).

Meanwhile, I'll be trying to figure out how to fix broken tests :sweat_smile:

LeXofLeviafan commented:
…It appears that tests are failing because vcr-cassettes are broken now. Not sure why they pass on `py37` though; it seems to be failing when I run it locally.

LeXofLeviafan commented on bukuserver/api.py:
```diff
@@ -158,45 +173,51 @@ def delete(self, rec_id: T.Union[int, None]):
+                form.process_data(request.get_json().get(str(rec_id)))
+            except (ValueError, TypeError):
+                return Response.INPUT_NOT_VALID()
+            kwargs_list.append({'index': rec_id,
```
Wasn't there something about `+= […]` being preferred syntax for pushing data into list, according to the style guide?

LeXofLeviafan commented on bukuserver/api.py:
```diff
@@ -158,45 +173,51 @@ def delete(self, rec_id: T.Union[int, None]):
+        if starting_id > ending_id or ending_id > max_id:
+            return Response.FAILURE()
+
+        kwargs_list = []
```
Something like `updates` would be less abstract/more on point as a name here.

LeXofLeviafan commented on bukuserver/api.py:
```diff
@@ -158,45 +173,51 @@ def delete(self, rec_id: T.Union[int, None]):
+            try:
+                form.process_data(request.get_json().get(str(rec_id)))
+            except (ValueError, TypeError):
+                return Response.INPUT_NOT_VALID()
```
…Maybe supply additional info about the error here, via metadata? (To make it easier to identify the cause of request rejection.)

LeXofLeviafan commented on bukuserver/forms.py:
```diff
@@ -1,26 +1,69 @@
+
+    def process_data(self, data):
+        tags = data.get('tags')
+        if tags and not isinstance(tags, list):
```
Technically `tags = ''` would pass here despite being invalid. And also `tags = [42]`, methinks?

May want to add these cases to tests, either way.

LeXofLeviafan commented on bukuserver/forms.py:
```diff
@@ -1,26 +1,69 @@
+            raise TypeError("List of tags expected.")
+        super().process(data=data)
+        if not self.validate():
+            raise ValueError("Input data not valid.")
```
Would it be possible to provide more information here? (E.g. which check failed)

2ynn commented:
Hi @LeXofLeviafan, thanks for your review and sorry again for disappearing :/
I just pushed some updates to my PR:
1. added `del_tags` boolean field to `ApiBookmarkRangeEditForm`. By default `del_tags` is set to `False`,
   meaning that the list of tags provided will be *appended* to the existing tags (i.e. `+` is prepended to the comma-separated list). Passing `del_tags: True` along with the request will instead prepend `-`.  `del_tags` can be set independently for each record.
2. added more detailed error messages

LeXofLeviafan commented on bukuserver/response.py:
```diff
@@ -1,17 +1,23 @@
+
+    @staticmethod
+    def bad_request(message: str):
+        json = {status: Response.FAILURE.status, message: message}  # pylint: disable=E0602
```
Er... maybe instead of disabling variable check, just use strings as dict keys? 😅

LeXofLeviafan reviewed (changes requested):
These and the nonexistent-variable one are causing the `flake8` check to fail.

The unit tests are failing on master as well, so I reported a bug regarding the error (#698).

LeXofLeviafan commented:
> The unit tests are failing on master as well, so I reported a bug regarding the error (#698).

Tests were fixed in upstream; your branch should be passing once you rebase and fix `flake8` errors.

LeXofLeviafan commented on bukuserver/api.py:
```diff
@@ -16,7 +16,7 @@
     from response import Response
     from forms import ApiBookmarkCreateForm, ApiBookmarkEditForm, ApiBookmarkRangeEditForm, ApiTagForm
-except ImportError:
+except (ImportError, ModuleNotFoundError):
```
[`ModuleNotFoundError` is a subclass of `ImportError`](https://docs.python.org/3.8/library/exceptions.html?highlight=modulenotfounderror#ModuleNotFoundError); this change doesn't actually have any effect. (You seem to have missed `During handling of the above exception, another exception occurred:` in the log, indicating that the actual issue happened _inside_ of the `except:` block.)

LeXofLeviafan commented on bukuserver/forms.py:
```diff
@@ -1,5 +1,6 @@
 """Forms module."""
 # pylint: disable=too-few-public-methods, missing-docstring
+from __future__ import annotations
```
Workarounds like these should have a comment to indicate specific purpose; i.e.:
```py
from __future__ import annotations  # Python <3.9 may need this to handle typehints like tuple[...]
```

LeXofLeviafan reviewed (changes requested):
…Incidentally, here's a quote from [the very first comment](https://github.com/jarun/buku/pull/676#issuecomment-1427288222):

> P.S. You may want to amend existing commits & force-push your changes instead of adding new commits; the repo owner prefers to avoid having multiple small commits (but he doesn't want to use squash-merge).

(Multiple commits can be squashed together using soft-reset + amend)

jarun commented:
@LeXofLeviafan please merge when this is ready.

LeXofLeviafan commented on bukuserver/response.py:
```diff
@@ -1,4 +1,47 @@
+    def json(self, data :dict = None) -> dict:
+        return dict(status=self.status, message=self.message, **data or {})  # pylint: disable=R1735
+
+    def __call__(self, *, data :dict = None):
```
…Come to think of it – why do you have such odd spacing for these particular typehints?
(`param :type` instead of `param: type`)

LeXofLeviafan commented on tests/test_server.py:
```diff
@@ -38,204 +52,240 @@ def test_home(client):
+
+    rd = client.put('/api/bookmarks/1/2', json={
+        1: {'tags': ['tag1 A', 'tag1 B', 'tag1 C']},
+        2: {'tags': ['tag2']}
```
Technically only string keys are valid in JSON dicts

LeXofLeviafan commented:
…Merging [as requested](https://github.com/jarun/buku/pull/676#issuecomment-1768879822).

LeXofLeviafan commented:
…Incidentally, I suggest referencing _the issue_ in commit name next time (not _the pull-request_) :sweat_smile:

Also [linking the pull-request to the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) is generally a good idea (happens automatically if its description is worded accordingly).
